docker-plugin/src/main/java/io/micronaut/gradle/docker/NativeImageDockerfile.java

```diff
@@ -457,16 +458,21 @@ private CopyFileInstruction toCopyResourceDirectoryInstruction(java.io.File reso
 
     protected List<String> buildActualCommandLine(Provider<String> executable,
                                                   DockerBuildStrategy buildStrategy,
-                                                  BaseImageForBuildStrategyResolver imageResolver) {
+                                                  BaseImageForBuildStrategyResolver imageResolver,
+                                                  ApplicationType applicationType) {
         NativeImageOptions options = newNativeImageOptions("actualDockerOptions");
         prepareNativeImageOptions(options);
         if (buildStrategy == DockerBuildStrategy.ORACLE_FUNCTION) {
             options.getMainClass().set("com.fnproject.fn.runtime.EntryPoint");
             options.getBuildArgs().add("--report-unsupported-elements-at-runtime");
         } else if (buildStrategy == DockerBuildStrategy.LAMBDA) {
             JavaApplication javaApplication = getProject().getExtensions().getByType(JavaApplication.class);
-            if (!options.getMainClass().isPresent()) {
-                options.getMainClass().set(javaApplication.getMainClass().orElse(DEFAULT_LAMBDA_RUNTIME_CLASS));
+            if (applicationType == ApplicationType.DEFAULT) {
```

Comment:
:x: I am not sure about this change: this is going to overwrite whatever is configured for `options.getMainClass()`, which we don't want.

Reply:
I researched more. The way it was designed was. 

# Type: Application.

For applications of type `Application`. Applications you use `@Controller` with.  

We want to use a `mainClass` for development and test and different `mainClass` for NativeImage. 

We want users to be able during development to able to run with `./gradlew run`. 

That it is way the generated apps have a regular `Application` class and we set it as `mainClass`. That it is why we add [`function-aws-api-proxy`](https://github.com/micronaut-projects/micronaut-aws/tree/master/function-aws-api-proxy) dependency as `developmentOnly` and also in the test classpath. 

When we package as a Native image we need to use `DEFAULT_LAMBDA_RUNTIME_CLASS`. 

# For Type Function 

We want to use whather the user set in the `application` plugin configuration. We don't want to override it. 



Actually, the w. For During development,  what we want to allow users to run their app with `./gradlew run` using a regular `Application.class`. 

Only use ``

We want users to be able to use the undocumented module `function-a

Reply:
But if you remove the `if (!options.getMainClass()` call, then any change to `graalvmNative { binaries { mainClass = ... } }` will be ignored. This would be inconsistent, no?

Reply:
I have changed it so that it is not breaking with 3.3.1 for DEFAULT applications: 

In 3.3.1 

```
} else if (buildStrategy == DockerBuildStrategy.LAMBDA) {
    JavaApplication javaApplication = getProject().getExtensions().getByType(JavaApplication.class);
    if (!javaApplication.getMainClass().isPresent()) {
        options.getMainClass().set("io.micronaut.function.aws.runtime.MicronautLambdaRuntime");
    }
    if (!options.getMainClass().isPresent()) {
        options.getMainClass().set("io.micronaut.function.aws.runtime.MicronautLambdaRuntime");
    }
}
```

With this change: 

```
} else if (buildStrategy == DockerBuildStrategy.LAMBDA) {
    JavaApplication javaApplication = getProject().getExtensions().getByType(JavaApplication.class);
    if (applicationType == ApplicationType.DEFAULT) {
        if (!javaApplication.getMainClass().isPresent()) {
            options.getMainClass().set("io.micronaut.function.aws.runtime.MicronautLambdaRuntime");
        }
        if (!options.getMainClass().isPresent()) {
           options.getMainClass().set("io.micronaut.function.aws.runtime.MicronautLambdaRuntime");
        }
    } else {
        if (!options.getMainClass().isPresent()) {
            options.getMainClass().set(javaApplication.getMainClass().orElse(DEFAULT_LAMBDA_RUNTIME_CLASS));
        }
    }
}
```